### DISCUSSION TITLE
Behavior and non-behavior

### DISCUSSION THREAD

**Discussion by cchannon** [microsoft/Power-Fx#818] [Q&A] [answered]
I keep getting stymied by these "Behavior function in a non-behavior property" errors as I am trying to build my own PFX implementation. In general, I get the idea: there are _properties_ that expect an output and not an operation. and there are _operations_ that expect actions (or behavior) and do not expect (or, at least, do not require) an output.

But I am having a hard time wrapping my head around how the Eval method is determining which expressions _are_ and _are not_ non-behavior properties. Sometimes I get this error on statements as simple as "Set(avariable, 1)" and I find it hard to imagine how that could be seen as a "non-behavior property". Is there a reference page somewhere that can help me understand this? The core documentation wiki reads kinda like a CS textbook (no offense!) so it is hard to figure out what I am looking for there.

**Comment by anderson-joyle** [comment, marked as answer]
A behavior function is a function that changes the value of a variable, or database table, or any other value that is somewhere previously stored. Most of Power Fx formulas (non-behavior) receive 0..N arguments, return a calculated value, and that's the end of it. The behavior function result, in the other hand, persists.
Currently there are mainly 4 behavior functions: [Set](https://learn.microsoft.com/en-us/power-platform/power-fx/reference/function-set), [Collect](https://learn.microsoft.com/en-us/power-platform/power-fx/reference/function-clear-collect-clearcollect), [Patch](https://learn.microsoft.com/en-us/power-platform/power-fx/reference/function-patch), and [Remove](https://learn.microsoft.com/en-us/power-platform/power-fx/reference/function-remove-removeif).
To enable the behavior functions, you got to explicitly "tell" the engine to. Please check the following test cases:
https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Interpreter.Tests/MutationFunctionsTests.cs

Take a closer look at these two lines:  

line 29
`private readonly ParserOptions _opts = new ParserOptions { AllowsSideEffects = true };`  

line 37  
`symbol.EnableMutationFunctions();`

**Reply by cchannon** [reply]
OK, I am tracking this, and yes, these changes definitely smooth out the wrinkles in my tests, but I don't actually _understand_ these changes yet. What does it actually mean to "Allow Side Effects"? And are "mutation functions" synonymous with Behavior functions, or is there a difference there?

**Reply by anderson-joyle** [reply]
Mutation is synonymous with Behavior.
"Allow Side Effects" means 2 things:

- Functions are allowed to change value from a source.
-  Allows formula chaining (semicolon operator) e.g. "**Set(a,10);a*2**"
